File: algorithm/code/TypeScript/stack/linkedliststack.test.ts
import LinkedListStack from './linkedliststack';

// 定义一个简单的类型用于测试
type TestDataType = number;

describe('LinkedListStack', () => {
    let stack: LinkedListStack<TestDataType>;

    beforeEach(() => {
        stack = new LinkedListStack<TestDataType>();
    });

    test('should be empty on initialization', () => {
        expect(stack.isEmpty()).toBe(true);
        expect(stack.size()).toBe(0);
    });

    test('should push elements onto the stack', () => {
        stack.push(1);
        stack.push(2);
        stack.push(3);

        expect(stack.isEmpty()).toBe(false);
        expect(stack.size()).toBe(3);
        expect(stack.top()).toBe(3);
    });

    test('should pop elements from the stack', () => {
        stack.push(1);
        stack.push(2);
        stack.push(3);

        expect(stack.pop()).toBe(3);
        expect(stack.size()).toBe(2);
        expect(stack.top()).toBe(2);
    });

    test('should return null when popping from an empty stack', () => {
        expect(stack.pop()).toBeNull();
    });

    test('should return the top element without removing it', () => {
        stack.push(1);
        expect(stack.top()).toBe(1);
        expect(stack.size()).toBe(1);
    });

    test('should clear the stack', () => {
        stack.push(1);
        stack.push(2);
        stack.clear();

        expect(stack.isEmpty()).toBe(true);
        expect(stack.size()).toBe(0);
    });

    test('should print the stack elements', () => {
        const consoleSpy = jest.spyOn(console, 'log');
        stack.push(1);
        stack.push(2);
        stack.push(3);

        stack.print();

        expect(consoleSpy).toHaveBeenCalledWith(3);
        expect(consoleSpy).toHaveBeenCalledWith(2);
        expect(consoleSpy).toHaveBeenCalledWith(1);

        consoleSpy.mockRestore(); // 清理 spy
    });
});

File: algorithm/code/Go/stack/linkedliststack_test.go
package stack

import (
	"fmt"
	"testing"
)

// 定义一个简单的类型，实现 Comparable 接口
type Int int

func (a Int) Compare(other Int) bool {
	return a == other
}

func TestLinkedListStack(t *testing.T) {
	stack := NewLinkedListStack[Int]()

	// 测试初始状态
	if !stack.IsEmpty() {
		t.Errorf("Expected stack to be empty")
	}
	if size := stack.Size(); size != 0 {
		t.Errorf("Expected size to be 0, got %d", size)
	}

	// 测试入栈
	stack.Push(1)
	stack.Push(2)
	stack.Push(3)

	// 测试栈的状态
	if stack.IsEmpty() {
		t.Errorf("Expected stack to not be empty")
	}
	if size := stack.Size(); size != 3 {
		t.Errorf("Expected size to be 3, got %d", size)
	}

	// 测试栈顶元素
	if top, ok := stack.Top(); !ok || top != 3 {
		t.Errorf("Expected top element to be 3, got %d", top)
	}

	// 测试出栈
	if popped, ok := stack.Pop(); !ok || popped != 3 {
		t.Errorf("Expected popped element to be 3, got %d", popped)
	}
	if size := stack.Size(); size != 2 {
		t.Errorf("Expected size to be 2 after pop, got %d", size)
	}

	// 测试再次出栈
	if popped, ok := stack.Pop(); !ok || popped != 2 {
		t.Errorf("Expected popped element to be 2, got %d", popped)
	}

	// 测试栈顶元素
	if top, ok := stack.Top(); !ok || top != 1 {
		t.Errorf("Expected top element to be 1, got %d", top)
	}

	// 测试清空栈
	stack.Clear()
	if !stack.IsEmpty() {
		t.Errorf("Expected stack to be empty after clear")
	}
	if size := stack.Size(); size != 0 {
		t.Errorf("Expected size to be 0 after clear, got %d", size)
	}

	// 测试从空栈出栈
	if popped, ok := stack.Pop(); ok || popped != 0 {
		t.Errorf("Expected popped element from empty stack to be 0 and ok to be false, got %d", popped)
	}

	// 测试栈的打印（可选）
	fmt.Println("Current stack state:")
	stack.Print()
}
